saleor/site/models.py

```diff
@@ -1,26 +1,83 @@
-from django.contrib.sites.models import _simple_domain_name_validator
+import threading
+
+from django.contrib.sites.models import SiteManager, Site
+from django.core.exceptions import ImproperlyConfigured
 from django.db import models
+from django.http.request import split_domain_port
 from django.utils.encoding import python_2_unicode_compatible
 from django.utils.translation import pgettext_lazy
 
 from . import AuthenticationBackends
 
 
+lock = threading.Lock()
+with lock:
+    THREADED_SITE_CACHE = {}
+
+
+def new_get_current(self, request=None):
+    from django.conf import settings
+    if getattr(settings, 'SITE_ID', ''):
+        site_id = settings.SITE_ID
+        if site_id not in THREADED_SITE_CACHE:
+            with lock:
+                site = self.prefetch_related('settings').filter(pk=site_id)[0]
+                THREADED_SITE_CACHE[site_id] = site
+        return THREADED_SITE_CACHE[site_id]
+    elif request:
+        host = request.get_host()
+        try:
+            # First attempt to look up the site by host with or without port.
+            if host not in THREADED_SITE_CACHE:
+                with lock:
+                    site = self.prefetch_related('settings').filter(
+                        domain__iexact=host)[0]
+                    THREADED_SITE_CACHE[host] = site
+            return THREADED_SITE_CACHE[host]
+        except Site.DoesNotExist:
+            # Fallback to looking up site after stripping port from the host.
+            domain, port = split_domain_port(host)
+            if domain not in THREADED_SITE_CACHE:
+                with lock:
+                    site = self.prefetch_related('settings').filter(
+                        domain__iexact=domain)[0]
+                    THREADED_SITE_CACHE[domain] = site
+        return THREADED_SITE_CACHE[domain]
+
+    raise ImproperlyConfigured(
+        "You're using the Django \"sites framework\" without having "
+        "set the SITE_ID setting. Create a site in your database and "
+        "set the SITE_ID setting or pass a request to "
+        "Site.objects.get_current() to fix this error."
+    )
+
+
+def new_clear_cache(self):
+    global THREADED_SITE_CACHE
+    with lock:
+        THREADED_SITE_CACHE = {}
+
+
+def new_get_by_natural_key(self, domain):
+    return self.prefetch_related('settings').filter(domain__iexact=domain)[0]
+
+
+SiteManager.get_current = new_get_current
+SiteManager.clear_cache = new_clear_cache
+SiteManager.get_by_natural_key = new_get_by_natural_key
```

Comment:
I also think all of the logic related to patching `django.contrib.sites` should be enclosed in separate file (example `patch_sites.py`) or clarity, and the three lines above that do the actual patching also could live in a function `patch_contrib_sites`. In this module we should have some commentary explaining why we're doing this.